File: src/main/java/com/raiffeisentesttask/accountingofsocks/aos/ExceptionAdvice.java
package com.raiffeisentesttask.accountingofsocks.aos;

import com.raiffeisentesttask.accountingofsocks.aos.entity.Response;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.web.bind.MissingServletRequestParameterException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

@ControllerAdvice
public class ExceptionAdvice {

    @ExceptionHandler({HttpMessageNotReadableException.class, MissingServletRequestParameterException.class})
    public ResponseEntity<Response> handleExceptionBadRequest(){
        Response response = new Response("HTTP 400 — параметры запроса отсутствуют или имеют некорректный формат");
        return new ResponseEntity<>(response, HttpStatus.BAD_REQUEST);
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<Response> handleExceptionServerError(){
        Response response = new Response("HTTP 500 — произошла ошибка, не зависящая от вызывающей стороны");
        return new ResponseEntity<>(response, HttpStatus.INTERNAL_SERVER_ERROR);
    }
}

File: core/src/main/kotlin/cib/interns/test/task/controller/ExceptionAdvice.kt
package cib.interns.test.task.controller

import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice
import javax.validation.ConstraintViolationException

@RestControllerAdvice
class ExceptionAdvice {

    @ExceptionHandler(ConstraintViolationException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    fun handle(ex: ConstraintViolationException) : List<Message> {

        val list = mutableListOf<Message>()

        ex.constraintViolations.iterator().forEach {
            list.add(Message("VALIDATION_EXCEPTION", it.propertyPath.toString(), it.message))
        }

        return list
    }

    data class Message(val type: String, val path: String, val message: String)
}